### PR TITLE
Automatically rename binaries in CI during Regular releases

### PR DESCRIPTION
**Pull request by DeeDeeG**
<details><summary><b>Requirements for Adding, Changing, or Removing a Feature</b> (from template, click to expand):</summary>

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must contribute a change that has been endorsed by the maintainer team. See details in the template below.
* The pull request must update the test suite to exercise the updated functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

</details>

### Issue or RFC Endorsed by <s>Atom's</s> Pulsar's Maintainers

We discussed it briefly in Discord...
https://discord.com/channels/992103415163396136/992109539346370661/1141120286473850970

### Description of the Change

Rename the produced binaries in Cirrus CI for the Regular release binary naming pseudo-convention<sup>[\[1\]](https://github.com/pulsar-edit/.github/pull/199)</sup><sup>[\[2\]](https://github.com/orgs/pulsar-edit/discussions/134)</sup>, automatically instead of having to do it manually.

Reads the last segment of the version string to determine whether the build is a Regular release build or something else (namely a "Rolling" style build).

This is a self-contained NodeJS script with no external dependencies, and should run on a wide array of Node versions, but tested on a recent Node 16.x release only.

### Alternate Designs

Do it manually, or different automation approach...? Hypothetically?

### Possible Drawbacks

Automating sets more stuff slightly in stone. It also adds convenience only insofar as it doesn't get in the way.

Automation is a double-edged sword, so I did try to make this script "less smart" (inspired by Mauricio's past complaints about "too smart" release tooling), by making us manually specify the prefix to add, for example with the prefix "Silicon.Mac.[binary name]", this doesn't try to infer what platform we're building on, the prefix is passed as an explicit argument to the script, written the right way per platform in the `.cirrus.yml`.

This does try to read the end of the version string and see if it's meant to be Rolling or Regular. I think "last segment >= 4 characters long = Rolling" is a fair heuristic for now, and it's what I ended up using here.

"A little smart" but "not too smart" is what I'm going for.

### Verification Process

Tested locally and in CI. Appears to work -- ~~runs to confirm a last minute fix to the version number heuristic for discerning "Rolling" vs "Regular" are still happening in CI, but were confirmed locally already.~~

UPDATE: After several fixes, this is working in CI as well as locally now.

Rolling-style version string in `package.json` (skips renaming the binaries, as intended): https://cirrus-ci.com/build/6586412099698688
Regular-style version string in `package.json` (renames the binaries successfully and tests pass as well as they did before, again as intended): https://cirrus-ci.com/build/6466216500396032

### Release Notes

N/A

### CODE REVIEW

**Review by confused-Techie — APPROVED** [review]
This looks great to me!

The next best thing is to try and automate the rest of the release pipeline lol.

But seriously rad changes, and if your tests are looking good locally, I say lets give it a shot on this release and merge

**Review by DeeDeeG — COMMENTED** [review]
So, in my mind this script should have been incapable of erroring out an otherwise passing CI run. _(I forgot to effectively ensure that in this quick implementation, though. Juggling a little too many things to remember to do that, maybe. So, probably good this was delayed until after the ongoing Regular release happening at the moment.)_

Which means I would have had to have used async or Promise based Node `fs` operations with error handling, for the scenarios where those can potentially error, or put just those things liable to error out in try/catch blocks with good error logging, or just put the whole script in a try/catch and at least print whatever the error was at that coarse level.

I should ideally come back and implement one of these approaches before we merge this. Worst case this script should do nothing, rather than error out and otherwise passing CI run, IMO. First do no harm, etc. Trying to reduce the "double edged sword" aspect of automation here.

tl;dr: **gotta add some try/catch to this.**

File: script/rename.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+const versionSegments = pulsarVersion.split('.');
+const lastSegment = versionSegments[versionSegments.length - 1];
+
+if (lastSegment.length > 4) {
+	// For example, '1-dev' is longer than 4 characters,
+	// and the format like '2023081600' from CI is longer than 4 characters.
+	// Either of those would indicate Rolling, not Regular.
+	console.log(`Based on the version string in package.json (${pulsarVersion}),`);
+	console.log('we are *not* preparing a Regular release, so *not* renaming any binaries.');
+	console.log('Exiting the binary renaming script.');
```
Comment (savetheclocktower) [inline script/rename.js:17]: One thing you can do in order to improve readability would be to explicitly exit here with

```js
process.exit();
```

so that you wouldn't have to put the rest of the script in an `else` conditional.

File: script/rename.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+} else {
+
+	// Warn about required prefix as first argument.
+	// If prefix was not provided, print warning and do nothing else, exiting "cleanly" with status 0.
+	if (typeof prefix !== "string" || prefix.length === 0) {
+		console.log("A prefix is required. Pass the desired prefix as the first argument to this script.");
+		console.log('Example usage: "node script/rename.js Windows"');
+		console.log('or: "node script/rename.js ARM.Linux"');
+		console.log("Exiting the binary renaming script.");
```
Comment (savetheclocktower) [inline script/rename.js:26]: You could do the same thing here

```js
process.exit();
```

and unwrap the _next_ `else` conditional.

**Review by savetheclocktower — COMMENTED** [review]
Just a couple of suggestions, neither one a blocker. Feel free to take the advice or ignore it.

File: script/rename.js (hunk 1 above)
Comment (DeeDeeG) [reply script/rename.js:17]: I read some stuff once in the NodeJS `process` API notes that has made me avoid `process.exit()` in Node scripts ever since.

https://nodejs.org/api/process.html#processexitcode

Not sure how real-world relevant it is, whether in general or to this script. But doing control flow to not schedule any work past "exit" conditions was my deliberate approach here.

I'm inclined to leave it, in case there is some async thing in this script that exits weirdly if we do `process.exit()`, not that I do much async in this _at all_. Probably okay to `process.exit()` in this script, but leaning away for aforementioned reasons.

Comment (savetheclocktower) [reply script/rename.js:17]: Eesh. Yeah, if it were my script I'd feel comfortable exiting early here considering lack of spooky async action, but I take your point.

**Comment by DeeDeeG** [comment]
Added a `try`/`catch` around the whole script. This handles any possible errors "gracefully" -- or at least exits with a clean exit status if anything _does_ go wrong, so as to never cause a build to fail from the JS side of things.

And finally fixed various issues with how I edited `.cirrus.yml` or other changes needed to work with the renamed binaries in the Cirrus config.

The CI run for the `exercise-rename-script` branch (https://cirrus-ci.com/build/6466216500396032) shows it _actually renaming binaries_ successfully, given enough retries for general (unrelated) CI flakiness. Whereas on this PR's branch, we still have a Rolling-style version string in `package.json` so the script rightly skips doing any renaming on this PR's branch.

<details><summary>Example output of the script, copy-pasted from real CI runs (click to expand):</summary>

## Windows

Rolling:

```
node script/rename.js "Windows"
Starting the binary renaming script...
Based on the version string in package.json (1.107.2023081700),
we are *not* preparing a Regular release, so *not* renaming any binaries.
Exiting the binary renaming script.
```

Regular:

```
node script/rename.js "Windows"
Starting the binary renaming script...
Based on the version string in package.json (1.107.999),
we *are* preparing a Regular release, so *renaming all binaries with prefixes*.
Prefix is: Windows
Existing filename is: Pulsar Setup 1.107.999.exe
Renaming file from "Pulsar Setup 1.107.999.exe" to "Windows.Pulsar.Setup.1.107.999.exe"
Existing filename is: Pulsar Setup 1.107.999.exe.blockmap
Renaming file from "Pulsar Setup 1.107.999.exe.blockmap" to "Windows.Pulsar.Setup.1.107.999.exe.blockmap"
Existing filename is: Pulsar-1.107.999-win.zip
Renaming file from "Pulsar-1.107.999-win.zip" to "Windows.Pulsar-1.107.999-win.zip"
```

### ARM Linux

Rolling:

```
node script/rename.js "ARM.Linux"
Starting the binary renaming script...
Based on the version string in package.json (1.107.2023081700),
we are *not* preparing a Regular release, so *not* renaming any binaries.
Exiting the binary renaming script.
```

Regular:

```
node script/rename.js "ARM.Linux"
Starting the binary renaming script...
Based on the version string in package.json (1.107.999),
we *are* preparing a Regular release, so *renaming all binaries with prefixes*.
Prefix is: ARM.Linux
Existing filename is: Pulsar-1.107.999-arm64.AppImage
Renaming file from "Pulsar-1.107.999-arm64.AppImage" to "ARM.Linux.Pulsar-1.107.999-arm64.AppImage"
Existing filename is: pulsar-1.107.999-arm64.tar.gz
Renaming file from "pulsar-1.107.999-arm64.tar.gz" to "ARM.Linux.pulsar-1.107.999-arm64.tar.gz"
Existing filename is: pulsar-1.107.999.aarch64.rpm
Renaming file from "pulsar-1.107.999.aarch64.rpm" to "ARM.Linux.pulsar-1.107.999.aarch64.rpm"
Existing filename is: pulsar_1.107.999_arm64.deb
Renaming file from "pulsar_1.107.999_arm64.deb" to "ARM.Linux.pulsar_1.107.999_arm64.deb"
```

### Apple Silicon Mac

Rolling:

```
export PATH="/opt/homebrew/bin:/opt/homebrew/opt/node@16/bin:$PATH"
node script/rename.js "Silicon.Mac"
Starting the binary renaming script...
Based on the version string in package.json (1.107.2023081623),
we are *not* preparing a Regular release, so *not* renaming any binaries.
Exiting the binary renaming script.
```

Regular:

```
export PATH="/opt/homebrew/bin:/opt/homebrew/opt/node@16/bin:$PATH"
node script/rename.js "Silicon.Mac"
Starting the binary renaming script...
Based on the version string in package.json (1.107.999),
we *are* preparing a Regular release, so *renaming all binaries with prefixes*.
Prefix is: Silicon.Mac
Existing filename is: Pulsar-1.107.999-arm64-mac.zip
Renaming file from "Pulsar-1.107.999-arm64-mac.zip" to "Silicon.Mac.Pulsar-1.107.999-arm64-mac.zip"
Existing filename is: Pulsar-1.107.999-arm64-mac.zip.blockmap
Renaming file from "Pulsar-1.107.999-arm64-mac.zip.blockmap" to "Silicon.Mac.Pulsar-1.107.999-arm64-mac.zip.blockmap"
Existing filename is: Pulsar-1.107.999-arm64.dmg
Renaming file from "Pulsar-1.107.999-arm64.dmg" to "Silicon.Mac.Pulsar-1.107.999-arm64.dmg"
Existing filename is: Pulsar-1.107.999-arm64.dmg.blockmap
Renaming file from "Pulsar-1.107.999-arm64.dmg.blockmap" to "Silicon.Mac.Pulsar-1.107.999-arm64.dmg.blockmap"
```

</details>

<details><summary>Example error handling (click to expand):</summary>

Script called incorrectly (no prefix given as first argument):

```console
% node script/rename.js 
Starting the binary renaming script...
A prefix is required. Pass the desired prefix as the first argument to this script.
Example usage: "node script/rename.js Windows"
or: "node script/rename.js ARM.Linux"
Exiting the binary renaming script.
```

With one of the Node `fs` operations edited to be incorrect and error out (strictly for purposes of demonstrating the `try`/`catch` error handling at the whole-script level):

```console
% node script/rename.js Intel.Mac                                                    
Starting the binary renaming script...
Based on the version string in package.json (1.107.999),
we *are* preparing a Regular release, so *renaming all binaries with prefixes*.
Prefix is: Intel.Mac
Existing filename is: Pulsar-1.101.0-dev-mac.zip
Renaming file from "Pulsar-1.101.0-dev-mac.zip" to "Intel.Mac.Pulsar-1.101.0-dev-mac.zip"
There was a problem during the renaming script!:
Error: ENOENT: no such file or directory, rename './binaries/Pulsar-1.101.0-dev-mac.zipxxx' -> './binaries/Intel.Mac.Pulsar-1.101.0-dev-mac.zip'
    at Object.renameSync (node:fs:1026:3)
    at /Users/[user]/pulsar/script/rename.js:49:8
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/[user]/pulsar/script/rename.js:39:13)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47 {
  errno: -2,
  syscall: 'rename',
  code: 'ENOENT',
  path: './binaries/Pulsar-1.101.0-dev-mac.zipxxx',
  dest: './binaries/Intel.Mac.Pulsar-1.101.0-dev-mac.zip'
}
% echo Exit status was: $?       
Exit status was: 0
```

</details>

**Comment by DeeDeeG** [comment]
So, this is ready for review again now that it's working as well as originally intended.

File: script/rename.js (hunk 1 above)
Comment (DeeDeeG) [reply script/rename.js:20]: I ended up doing it after all. I think @confused-Techie's "upload binaries to the Rolling repo" script is maybe modeled after this one a tiny bit, and it uses `process.exit()`, and everything is fine there, so I think I can relax a bit and do it here as well.

**Comment by DeeDeeG** [comment]
Okay, I updated this PR in the wake of the big "move most binary building from Cirrus to GitHub Actions" PR.

(I also used `process.exit()` after all, as @savetheclocktower  suggested. It seems to work just fine.)

**Ready for, and would appreciate, any additional review.**

_Tangentially related CI automation wishlist item: With this in place, it'd be real neat if we can get another "upload" script, but making a draft release for Regular with all the binaries, instead of an instant release for Rolling. And integrate it on the GitHub Actions "build binary" workflow and in the Cirrus config, if the `regular-release` PR label is set, I think was the label name we'd be using._

**Comment by confused-Techie** [comment]
Overall this looks amazing! The last thing I would actually request on this, is in the final `catch` statement, if we could add a `process.exit(1);` to ensure we get the failure status in CI, just in case.

But overall super glad to see these changes unified within the new CI situation, which again, sorry that blindsided this PR

**Comment by DeeDeeG** [comment]
@confused-Techie I would be willing to add `process.exit(1)` _if_ we set `allow_failure` or whatever it is on the job. Reason being, we might want to still have the binaries and be able to manually rename them, rather than having the whole build-test-publish workflow grind to a halt?

Renaming the binaries for Regular is a convenience utility to me, IMO, not essential. I'd rather be able to salvage the rest of the perfectly good run and use my own manual judgment whether to do a full re-run trying to fix the rename script or else just rename the binaries myself and fix the renaming script on a later run. If the rename script ever flakes, as a fluke we couldn't do anything about besides ignore that one time, then you'd just ignore it and move on.

**Comment by confused-Techie** [comment]
Ya know, @DeeDeeG you make a super valid point.

This is a convenience function, not mission critical. I was analyzing this incorrectly previously, and forgot that important tidbit.

Which in that case, ignore my previous suggestion, this looks great!

**Review by confused-Techie — APPROVED** [review]
LGTM! Lets get it merged
